central/metrics/aggregator/common/tracker_config.go

```diff
@@ -0,0 +1,165 @@
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/search"
+	"github.com/stackrox/rox/pkg/sync"
+)
+
+// TrackerConfig wraps various pieces of configuration required for tracking
+// various metrics.
+type TrackerConfig[Finding Countable] struct {
+	category    string
+	description string
+	labelOrder  map[Label]int
+	getters     map[Label]func(Finding) string
+	generator   FindingGenerator[Finding]
+	gauge       func(string, prometheus.Labels, int)
+
+	// metricsConfig can be changed with an API call.
+	metricsConfig    MetricsConfiguration
+	metricsConfigMux sync.RWMutex
+	query            *v1.Query
+
+	ticker *time.Ticker
+}
+
+type Tracker interface {
+	Run(context.Context)
+	Reconfigure(context.Context, *prometheus.Registry, string, map[string]*storage.PrometheusMetricsConfig_Labels, time.Duration) error
+}
+
+func makeLabelOrderMap[Finding Countable](getters []LabelGetter[Finding]) map[Label]int {
+	result := make(map[Label]int, len(getters))
+	for i, getter := range getters {
+		result[getter.Label] = i + 1
+	}
+	return result
+}
+
+func makeGettersMap[Finding Countable](getters []LabelGetter[Finding]) map[Label]func(Finding) string {
+	result := make(map[Label]func(Finding) string, len(getters))
+	for _, getter := range getters {
+		result[getter.Label] = getter.Getter
+	}
+	return result
+}
+
+// MakeTrackerConfig initializes a tracker configuration without any period or metrics configuration.
+// Call Reconfigure to configure the period and the metrics.
+func MakeTrackerConfig[Finding Countable](category, description string,
+	getters []LabelGetter[Finding], generator FindingGenerator[Finding],
+	gauge func(string, prometheus.Labels, int),
+) *TrackerConfig[Finding] {
+	return &TrackerConfig[Finding]{
+		category:    category,
+		description: description,
+		labelOrder:  makeLabelOrderMap(getters),
+		getters:     makeGettersMap(getters),
+		generator:   generator,
+		gauge:       gauge,
+	}
+}
+
+func (tc *TrackerConfig[Finding]) Reconfigure(ctx context.Context, registry *prometheus.Registry, filter string, cfg map[string]*storage.PrometheusMetricsConfig_Labels, period time.Duration) error {
+	mcfg, err := parseMetricLabels(cfg, tc.labelOrder)
+	if err != nil {
+		return err
+	}
+	q, err := search.ParseQuery(filter, search.MatchAllIfEmpty())
+	if err != nil {
+		return err
+	}
+	tc.SetMetricsConfiguration(q, mcfg)
+	// Force track only on reconfiguration, not on the initial tracker creation.
+	if tc.setPeriod(period) {
+		tc.track(ctx)
+	}
+	return tc.registerMetrics(registry, period)
+}
+
+// setPeriod returns true if tracker period has been reconfigured.
+func (tc *TrackerConfig[Finding]) setPeriod(period time.Duration) bool {
+	tc.metricsConfigMux.Lock()
+	defer tc.metricsConfigMux.Unlock()
+	if period > 0 {
+		if tc.ticker == nil {
+			tc.ticker = time.NewTicker(period)
+			return false
+		}
+		tc.ticker.Reset(period)
+		return true
+	}
+	if tc.ticker != nil {
+		tc.ticker.Stop()
+	}
+	return false
+}
```

Comment:
I find it overly complex. What the function does is not what the comment say. It should be reduced to `tc.ticker.Reset(period)`.